### ISSUE TITLE
Coverage builder is failing or flaking (Log in goroutine after TestConnClosed has completed)

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#1132]
@namansood reports that the coverage builder is returning spurious errors for a change of his:

```
PASS
coverage: 67.7% of statements
ok  	tailscale.com/wgengine/filter	0.104s
panic: Log in goroutine after TestConnClosed has completed

goroutine 884 [running]:
testing.(*common).logDepth(0xc000355080, 0xc0000780c0, 0x57, 0x3)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:755 +0x7ee
testing.(*common).log(...)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:737
testing.(*common).Logf(0xc000355080, 0xc0000300a0, 0x42, 0xc0001f6060, 0x2, 0x2)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:783 +0x90
tailscale.com/types/logger.WithPrefix.func1(0xc8cd0c, 0x3b, 0xc0001f6060, 0x2, 0x2)
	/home/runner/work/tailscale/tailscale/types/logger/logger.go:33 +0xb0
tailscale.com/wgengine/magicsock.(*discoEndpoint).sendPingsLocked(0xc0005d8000, 0xbff84ab9c4a93bc9, 0x1776427fa, 0x1047a00, 0x1)
	/home/runner/work/tailscale/tailscale/wgengine/magicsock/magicsock.go:3237 +0x6b5
tailscale.com/wgengine/magicsock.(*discoEndpoint).heartbeat(0xc0005d8000)
	/home/runner/work/tailscale/tailscale/wgengine/magicsock/magicsock.go:3049 +0x814
created by time.goFunc
	/opt/hostedtoolcache/go/1.15.6/x64/src/time/sleep.go:167 +0x52
exit status 2
FAIL	tailscale.com/wgengine/magicsock	6.316s
?   	tailscale.com/wgengine/monitor	[no test files]
?   	tailscale.com/wgengine/netstack	[no test files]
PASS
coverage: 54.1% of statements
ok  	tailscale.com/wgengine/router	0.077s
?   	tailscale.com/wgengine/router/dns	[no test files]
goos: linux
goarch: amd64
pkg: tailscale.com/wgengine/tsdns
BenchmarkFull/forward-2         	       1	     95810 ns/op
BenchmarkFull/reverse-2         	       1	    172117 ns/op
BenchmarkFull/delegated-2       	       1	    418841 ns/op
PASS
```

**Comment by DentonGentry** [comment]
Ah: TestConnClosed as it existed yesterday had a race condition. Last night I merged https://github.com/tailscale/tailscale/pull/1116 to revert it, and https://github.com/tailscale/tailscale/pull/1126 is out for review now with a revamped implementation to avoid the race.

I think your client branched from main at the point that it contained the flakey TestConnClosed. If you can git pull and rebase to the current main, it will remove that test case and you shouldn't run into the problem.

https://github.com/tailscale/tailscale/pull/1126 will go in later, and can run `go test -race -count=10000` without error so I think it avoids the race of the original test.

**Comment by josharian** [comment]
I think this is unrelated to #1116/#1126, and that if we bisected it we would discover that it has the same root cause as 18471a8792b74826672f0b0e2f2ab01c09cd1dfe. I'll investigate and fix.

**Comment by josharian** [comment]
I was partly wrong. Bisected to 2e9728023b49a10d6528b6cceff007339e775049, which is a new code-coverage inspired test. Investigating...

**Comment by josharian** [comment]
Most of this was fixed by #1135. I believe that the only remaining fix here is #1139.

**Comment by josharian** [comment]
#1139 is in. We're working on other stability issues, but this one is done.
